### PR TITLE
Ignore key events on edit control on web platform (#52656)

### PR DESCRIPTION
**Pull request by cjng96**
## Description

*Ignore key event from web platform to prevent twice key action by one key press.
For now It's done on Flutter Engine's keyboard.dart code. but I move that code here since we can use a key event via RawKeyListener on web platform(It's working well on Android)*

After this pull-request is merged, I will issue new pull-request on Flutter Engine that allowing to send keyEvent which occurs on input element with web.

## Related Issues

#52656

#44133

## Tests

I added the following tests:

editable_test.dart / ignore key event from web platform

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart]: This comment is unclear, is the key down only apply to arrow key?

File: packages/flutter/lib/src/rendering/editable.dart
Comment (cjng96) [reply packages/flutter/lib/src/rendering/editable.dart]: Good!. Thank you. :)

File: packages/flutter/lib/src/rendering/editable.dart
Comment (cjng96) [reply packages/flutter/lib/src/rendering/editable.dart]: This code replaces flutter engine's pull-request #13699. so ignore all keyEvent from input control on web.
Arrow key(_handleMovement function) and delete key(_handleDelete function) should be ignored. 

but I didn't check the keys in _handleShortcuts(it handles ctrl+c, ctrl+v, ctrl+x and ctrl+a) yet. I am going to check those keys should be ignored or not. then renew pull-request. Thank you.

Comment (cjng96) [reply packages/flutter/lib/src/rendering/editable.dart]: Ctrl+c, ctrl+v keys are working well. so I modify the code to ignore movement keys and delete key only.
Thank you.

File: packages/flutter/lib/src/rendering/editable.dart
Comment (cjng96) [reply packages/flutter/lib/src/rendering/editable.dart]: As you suggested, I tried to use the ServicesBinding class.

It seems to be working if I use testWidgets and EditableText for pumpWidget.
I failed to find a other solution to use both RenderEditable and ServicesBinding together without pumpWidget.

This test code is working well. but I am not sure if this is correct implementation under flutter testing rule. :(

Comment (chunhtai) [reply packages/flutter/lib/src/rendering/editable.dart]: Sorry I forgot to mention you will need TestWidgetsFlutterBinding.ensureInitialized(); 
See
https://github.com/flutter/flutter/blob/master/packages/flutter/test/services/platform_channel_test.dart

File: packages/flutter/lib/src/rendering/editable.dart
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/editable.dart]: We -> we

File: packages/flutter/test/rendering/editable_test.dart
Comment (chunhtai) [inline packages/flutter/test/rendering/editable_test.dart]: remove this?

File: packages/flutter/test/rendering/editable_test.dart
Comment (chunhtai) [inline packages/flutter/test/rendering/editable_test.dart]: I am surprised that uses `TestWidgetsFlutterBinding.ensureInitialized();` still requires testWidget. what was the issue you were facing? If this is necessary, this test should be put in editable_text_test.dart

File: packages/flutter_test/lib/src/event_simulation.dart
Comment (chunhtai) [inline packages/flutter_test/lib/src/event_simulation.dart]: just return code here and return null at line 115?

File: packages/flutter_test/lib/src/event_simulation.dart
Comment (chunhtai) [inline packages/flutter_test/lib/src/event_simulation.dart]: lets follow the same style here ```final int scanCode = platform == 'web' ? -1 : _getScanCode(physicalKey, platform);```

**Review by chunhtai — COMMENTED** [review]
added some comments, also cc @mdebbar

File: packages/flutter/lib/src/rendering/editable.dart
Comment (mdebbar) [inline packages/flutter/lib/src/rendering/editable.dart]: The reason we decided to let the browser handle all key events for us is because it's difficult to hardcode the handling of all keys.

For example, movement keys and delete are not the only ones we should ignore here. What about backspace, home, end, page-up, page-down, enter, etc?

Could you explain your use-case a bit more? Like, why do you need to listen to keys when the user is editing a text field?

Comment (cjng96) [reply packages/flutter/lib/src/rendering/editable.dart]: Hello thank you for your comment, :)

First, regarding the keys we should ignore on control,
**Duplicated operation on web is related to movement keys and delete key only** because it's handled in this _handleKeyEvent() function only.
Other keys(backspace, home, end, page-up...etc) are not handled here thus the duplicated operation doesn't happen. At least as far as I know. if there is anything I missd, please let me know.

Use-case for me is,
1. I hope to use ctrl+c, ctrl+v for copy and paste.
2. I need short-cut key. For example I popup dialog when the user press short-cut key on edit control.
3. I want to make custom rich editor control in future. I need receive key event on web for it. It's possible on Android or PC platform.

File: packages/flutter_test/lib/src/event_simulation.dart
Comment (cjng96) [reply packages/flutter_test/lib/src/event_simulation.dart]: This code is copied from _getKeyCode() function, this style is there. so I follow same style.

Anyway I think the style you suggest is better too. :)

File: packages/flutter_test/lib/src/event_simulation.dart
Comment (cjng96) [reply packages/flutter_test/lib/src/event_simulation.dart]: Good. OK. :)

File: packages/flutter/test/rendering/editable_test.dart
Comment (cjng96) [reply packages/flutter/test/rendering/editable_test.dart]: Oh sorry,

File: packages/flutter/test/rendering/editable_test.dart
Comment (cjng96) [reply packages/flutter/test/rendering/editable_test.dart]: Without testWidgets and pumpWidget, I could't use SystemBinding. :(
I will paste error message and sample code I tried because this can be user error(I'm newbie on Flutter).

I don't like this code which uses EditableText here. It's good if I can implement without testWidgets and pumpWidget function. and in other case, I will move this test code to editable_text_test.dart as you said.

Thank you.

Comment (cjng96) [reply packages/flutter/test/rendering/editable_test.dart]: > I am surprised that uses `TestWidgetsFlutterBinding.ensureInitialized();` still requires testWidget. what was the issue you were facing? If this is necessary, this test should be put in editable_text_test.dart

Sorry today I succeed to make test code without testWidgets and pumpWidget functions as you said. maybe I was a little bit confused about test framework. :(
I updated my pull-request. Thank you for your comments. :)

File: packages/flutter/test/rendering/editable_test.dart
Comment (chunhtai) [inline packages/flutter/test/rendering/editable_test.dart]: this can just be 
```dart
await simulateKeyDownEvent(LogicalKeyboardKey.arrowRight, platform: 'web');
```

File: packages/flutter/test/rendering/editable_test.dart
Comment (chunhtai) [inline packages/flutter/test/rendering/editable_test.dart]: same here

File: packages/flutter/test/rendering/editable_test.dart
Comment (cjng96) [reply packages/flutter/test/rendering/editable_test.dart]: Oh... I didn't know there is that global function. Thank you for shorten code.

**Review by chunhtai — COMMENTED** [review]
I don't have anything else, it lgtm if it also lgt @mdebbar

File: packages/flutter/lib/src/rendering/editable.dart
Comment (mdebbar) [inline packages/flutter/lib/src/rendering/editable.dart]: There's no need to check the type of the event. You should use the available constant `kIsWeb`.

And I would put this check in the beginning of the method to avoid doing unnecessary work like collapsing synonyms, etc.

**Review by mdebbar — COMMENTED** [review]
Could you also share the engine PR so we can review both at the same time?

File: packages/flutter/lib/src/rendering/editable.dart
Comment (mdebbar) [reply packages/flutter/lib/src/rendering/editable.dart]: Sorry for the delay. Yeah, those use-cases make sense.

**Comment by mdebbar** [comment]
Overall, the PR looks good to me at this point (except the `isWeb` check that I commented about). But I'll wait for the engine PR so we can review both and make sure they will work well together.

File: packages/flutter/lib/src/rendering/editable.dart
Comment (cjng96) [reply packages/flutter/lib/src/rendering/editable.dart]: Thank you. I modify the source to use kIsWeb. so I also modify test function. I activate my test function on web platform but omit other test functions in editable_test.dart on web. :(
I'm not sure if it's rightapproach. please confirm new pull-request.

**Comment by goderbauer** [comment]
This PR is blocked on https://github.com/flutter/engine/pull/17242.

**Review by mdebbar — APPROVED** [review]
LGTM

**Comment by cjng96** [comment]
I wonder when this pull-request is merged.

**Comment by mdebbar** [comment]
@cjng96 I can merge the PRs. I think the framework PR should be merged first, correct?

**Comment by cjng96** [comment]
> 
> 
> @cjng96 I can merge the PRs. I think the framework PR should be merged first, correct?

Yes, framework PR should be merged first. Thank you.

**Pull request merged**
